File: functions/scripts/python/tts_core.py
import os
from openai import OpenAI
from dotenv import load_dotenv
from schemas import Role

def get_voice_for_character(character):
    """Map character types to appropriate voices."""
    character_voices : Dict[Role, str] = {
        Role.NARRATOR: 'fable',  # Warm, engaging narrator voice
        Role.CHILD: 'nova',      # Youthful, bright voice
        Role.ELDER: 'onyx',      # Deep, authoritative voice
        Role.FAE: 'shimmer',   # Light, whimsical voice
        Role.HERO: 'alloy',      # Balanced, confident voice
        Role.VILLAIN: 'ash',     # Dramatic, resonant voice
        Role.SAGE: 'sage',
        Role.SIDEKICK: 'coral'
    }
    return character_voices.get(character.lower(), 'fable')  # Default to fable if character not found

def generate_speech_from_text(text, character="narrator", model="tts-1"):
    """Generate speech from text using OpenAI's TTS API."""
    # Load environment variables from two directories up
    env_path = os.path.join(os.path.dirname(__file__), '..', '..', '.env')
    load_dotenv(env_path)
    
    # Initialize OpenAI client
    client = OpenAI(api_key=os.getenv('OPENAI_API_KEY'))
    
    try:
        voice = get_voice_for_character(character)
        
        # Generate speech with streaming response
        response = client.audio.speech.create(
            model=model,
            voice=voice,
            input=text
        )
        
        return response
        
    except Exception as e:
        print(f"Error generating speech: {str(e)}")
        raise 

File: functions/src/tts_core.ts
import {OpenAI} from 'openai';
import * as dotenv from 'dotenv';
import * as path from 'path';
import {Role} from './schemas.js';

// OpenAI's supported voice types
type OpenAIVoice = 'alloy' | 'ash' | 'coral' | 'echo' | 'fable' | 'onyx' | 'nova' | 'sage' | 'shimmer';

// Type for the character voice mapping
type CharacterVoices = {
  [key in Role]: OpenAIVoice;
};

const characterVoices: CharacterVoices = {
  narrator: 'fable',  // Warm, engaging narrator voice
  child: 'nova',      // Youthful, bright voice
  elder: 'onyx',      // Deep, authoritative voice
  fairy: 'shimmer',   // Light, whimsical voice
  hero: 'alloy',      // Balanced, confident voice
  villain: 'ash',     // Dramatic, resonant voice
  sage: 'sage',
  sidekick: 'coral'
};

/**
 * Maps character types to appropriate voices
 * @param character - The character role
 * @returns The corresponding voice ID
 */
export function getVoiceForCharacter(character: Role): OpenAIVoice {
  return characterVoices[character] || 'fable';  // Default to fable if character not found
}

type TTSModel = 'tts-1' | 'tts-1-hd';

/**
 * Generate speech from text using OpenAI's TTS API
 * @param text - The text to convert to speech
 * @param character - The character role (defaults to narrator)
 * @param model - The TTS model to use
 * @returns Promise containing the audio response as an ArrayBuffer
 */
export async function generateSpeechFromText(
  text: string,
  character: Role = 'narrator',
  model: TTSModel = 'tts-1'
): Promise<ArrayBuffer> {
  // Load environment variables from two directories up
  const envPath = path.join(__dirname, '..', '.env');
  dotenv.config({ path: envPath });

  // Initialize OpenAI client
  const client = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY
  });

  try {
    const voice = getVoiceForCharacter(character);

    // Generate speech
    const response = await client.audio.speech.create({
      model,
      voice,
      input: text,
      response_format: 'mp3'
    });

    // Convert the response to ArrayBuffer
    const arrayBuffer = await response.arrayBuffer();
    return arrayBuffer;
  } catch (error) {
    console.error('Error generating speech:', error);
    throw error;
  }
} 